### DISCUSSION TITLE
Proxy seems to be not working

### DISCUSSION THREAD

**Discussion by claro-lukas** [MRColorR/money4band#29] [Q&A] [answered]
I tried to use proxies in .env file, like this:

PEER2PROFIT_HTTP_PROXY=http://IP:PORT
PEER2PROFIT_HTTPS_PROXY=http://IP:PORT

However proxies seems to be not working because PEER2PROFIT in the devices list still show real computer IP instead of proxy IP.

Is it a bug or proxies don't work with all apps?

**Comment by MRColorR** [comment]
Hello, and thank you for reaching out :) 
Have you added the proxy configuration manually? If so, you need to rename the placeholder flags `PEER2PROFIT_HTTP_PROXY`  to `HTTP_PROXY` (same for HTTPS) and recreate the containers.
If unsure all the settings can be easily configured using the setup scripts provided. Feel free to give them a try. 
If you used the automatic script, it could be a bug or regression in which case I need further details to hunt down the issue and fix it.

If you want to manually configure advanced options, please read both the project wiki and the docker compose documentation for more details and options.

**Reply by bass-77** [reply]
Hi MRColorR:)
using v1.31.
Trying to make new .env.
runme.sh
3) Setup .env file 
.. the same proxies for all..

getting .env file which contains:
PEER2PROFIT_HTTP_PROXY=http://myproxyip:myproxyport
PEER2PROFIT_HTTPS_PROXY=http://myproxyip:myproxyport

runme.sh
4) Start apps stack

No errors, containers start. The result is as falcodeveloper describes.
Peer2profit network panel shows the host IP with money4band dockers running, not proxy IP.

If I manually edit .env
PEER2PROFIT_HTTP_PROXY=http://myproxyip:myproxyport
PEER2PROFIT_HTTPS_PROXY=http://myproxyip:myproxyport
into
HTTP_PROXY=http://myproxyip:myproxyport
HTTPS_PROXY=http://myproxyip:myproxyport
and run runme.sh I get 
4) Start apps stack
WARN[0000] The "PEER2PROFIT_HTTP_PROXY" variable is not set. Defaulting to a blank string.
WARN[0000] The "PEER2PROFIT_HTTPS_PROXY" variable is not set. Defaulting to a blank string.

Could you please tell what I'm doing wrong. Thank you.

btw, the small feature request for multiple instances set up, it complains about existing container already running. Currently I have to edit docker-compose.yaml manually to change all container names.

**Comment by MRColorR** [comment]
Hello, something is wrong all the errors are related to an incomplete env or compose file (not your fault, maybe a bad script run). The output after running the option 3 of the setup should manage and create a different stack automatically to avoid container's name conflict. Also, it should change all the placeholder variable to the correct real ones.

on which system architecture and OS  are you two running the project ? i'll try to help you to get rid of this issue 
is the `sed` command recognized by your terminal ?
Is your project installation clean ? Please if possible use option 6 and 7 to reset env and compose file then run option 3 to reconfigure again

Edit: I just ran the setup script again and it works as expected. All values are successfully passed to variables. I have to take a cloud certification exam tomorrow, if the problem persists, I will come back to help you in 2 days.

**Reply by bass-77** [reply]
MRColoR, thank you a lot for your support.
There is some information about the system. Installation is fresh.

# lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 22.04.1 LTS
Release:        22.04
Codename:       jammy

# cat /etc/debian_version
bookworm/sid

# uname -a
Linux myhost 5.15.0-57-generic #63-Ubuntu SMP Thu Nov 24 13:43:17 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux

# sed --version
sed (GNU sed) 4.8
Packaged by Debian

# docker version
Client: Docker Engine - Community
 Version:           20.10.22
 API version:       1.41
 Go version:        go1.18.9
 Git commit:        3a2c30b
 Built:             Thu Dec 15 22:28:04 2022
 OS/Arch:           linux/amd64
 Context:           default
 Experimental:      true

Server: Docker Engine - Community
 Engine:
  Version:          20.10.22
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.18.9
  Git commit:       42c8b31
  Built:            Thu Dec 15 22:25:49 2022
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.6.15
  GitCommit:        5b842e528e99d4d4c1686467debf2bd4b88ecd86
 runc:
  Version:          1.1.4
  GitCommit:        v1.1.4-0-g5fd4c4d
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0

Trying to run project once more.
Stopping, removing containers:
docker stop $(docker ps -aq)
docker rm $(docker ps -aq)

bash ./runme.sh
6) Reset .env File
7) Reset docker-compose.yaml file

3) Setup .env file
Entering proxies, emails passwords..
No errors.
When setup reaches Bitping portion there is quite strange message:

/////
Go to BITPING | https://app.bitping.com?r=qm7mIuX3 and register
When done, press enter to continue

First of all make sure that you have already run the runme.sh script and that you have registered to the various apps with the links provided.
if you have already performed these actions, press any key to continue
/////

May be it is ok..
bitping docker starts, I don't have Bitping account, so I press ctrl+c.
Setup asks about updates, I choose N, and it finishes:

/////
Noted: all updates will be applied automatically and silently
env file setup complete.
Press any key to go back to the menu
/////

4) Start apps stack

I'm getting few errors, but setup continues :
--------------
This menu item will launch all the apps using the configured .env file and the docker-compose.yaml file (Docker must be already installed and running)
Do you wish to proceed Y/N?  y
WARN[0000] The "msyourIPROYALPAWNSPw" variable is not set. Defaulting to a blank string.
WARN[0000] The "msyourIPROYALPAWNSPw" variable is not set. Defaulting to a blank string.
WARN[0000] The "msyourIPROYALPAWNSPw" variable is not set. Defaulting to a blank string.
[+] Running 12/12
 ⠿ Network money4band_29111_default        Created                                                                                0.0s
 ⠿ Volume "money4band_29111_earnapp-data"  Created                                                                                0.0s
 ⠿ Container traffmonetizer                Started                                                                                1.3s
 ⠿ Container iproyal                       Started                                                                                1.3s
 ⠿ Container repocket                      Started                                                                                1.0s
 ⠿ Container earnapp                       Started                                                                                1.2s
 ⠿ Container peer2profit                   Started                                                                                1.1s
 ⠿ Container bitping                       Started                                                                                1.3s
 ⠿ Container webdashboard                  Started                                                                                1.4s
 ⠿ Container watchtower                    Started                                                                                1.3s
 ⠿ Container honeygain                     Started                                                                                1.0s
 ⠿ Container packetstream                  Started                                                                                0.7s
All Apps started you can visit the web dashboard on http://localhost:8081/ . If not already done use the previously generated earnapp node URL to add your device in your earnapp dashboard. Check the README file for more details.
Now press enter to go back to the menu
--------------
Containers are running, except
bitping, repocket, iproyal restarting..

Content of .env file (I suspect some characters should be escaped in passwords, I see wrong password line when it contains &):
# cat .env
######################################################################
# File name : .env
# Author : MRColor
#
# Environment used within docker-compose file
# WARNING : This file will contains sensitive datas such as login
#           and/or password.
#           Prevent access from unauthorized users on this machine.
######################################################################
#
#FIRST OF ALL PLEASE REGISTER ON THE PLATFORMS USING THIS LINKS, YOU'LL NEED TO ENTER SOME DATA BELOW:
# 1) Earnapp | https://earnapp.com/i/3zulx7k
# 2) HoneyGain | https://r.honeygain.me/MINDL15721
# 3) IPROYAL | https://pawns.app?r=MiNe
# 4) PACKETSTREAM | https://packetstream.io/?psr=3zSD
# 5) PEER2PROFIT | https://p2pr.me/165849012262da8d0aa13c8
# 6) TRAFFMONETIZER | https://traffmonetizer.com/?aff=366499
# 7) REPOCKET | https://link.repocket.co/hr8i
# 8) BITPING | https://app.bitping.com?r=qm7mIuX3
#
######################################################################
## PROJECT NAME
COMPOSE_PROJECT_NAME=Money4Band_XXXXX

## MACHINE NAME (for display in apps' dashboard if the app permits this)
DEVICE_NAME=m4b00

## EARNAPP
# Your Earnapp device id : ! must start with sdk-node-<md5sum>
# <md5sum> can be done this way : echo "your random string" | md5sum  (get the 32 char result)
# or just paste your device_name or another string to https://www.md5hashgenerator.com/ and copy the result in the device UUID after the sdk-node- part
# or you can generate one directly using the full privileged image for earnapp
# A few time after starting it successfully : register your node on earnapp dashboard : https://earnapp.com/r/<your device id>
EARNAPP_DEVICE_UUID=sdk-node-XXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
EARNAPP_HTTP_PROXY=http://XXXXXXXXXX:XXXX
EARNAPP_HTTPS_PROXY=http://XXXXXXXXXX:XXXX


## HONEYGAIN
# Your Honeygain account email
# Your Honeygain account password
HONEYGAIN_EMAIL=XXX@XXX.XXX
HONEYGAIN_PASSWD=XXXXXXXXXXXX
HONEYGAIN_HTTP_PROXY=http://XXXXXXXXXX:XXXX
HONEYGAIN_HTTPS_PROXY=http://XXXXXXXXXX:XXXX

## IPROYAL
# Your IproyalPawns account email
# Your IproyalPawns account password
IPROYALPAWNS_EMAIL=XXX@XXX.XXX
IPROYALPAWNS_PASSWD=XXXXXXXXXmsyourIPROYALPAWNSPw^
IPROYALPAWNS_HTTP_PROXY=http://XXXXXXXXXX:XXXX
IPROYALPAWNS_HTTPS_PROXY=http://XXXXXXXXXX:XXXX

## PEER2PROFIT
# Your Peer2profit account email
PEER2PROFIT_EMAIL=XXX@XXX.XXX
PEER2PROFIT_HTTP_PROXY=http://XXXXXXXXXX:XXXX
PEER2PROFIT_HTTPS_PROXY=http://XXXXXXXXXX:XXXX

## PACKETSTREAM
# Your packetstream CID (to retrieve in your packetstream dashboard : Section : download -> Looking for linux app -> docker command : -e CID=xxxx)
PACKETSTREAM_CID=XXX
PACKETSTREAM_HTTP_PROXY=http://XXXXXXXXXX:XXXX
PACKETSTREAM_HTTPS_PROXY=http://XXXXXXXXXX:XXXX

## TRAFFMONETIZER
# Your traffmonetizer token (to retrieve in your traffmonetizer dashboard : Section : dashboard -> Your application token
TRAFFMONETIZER_TOKEN=yourTRAFFMONETIZERToken
TRAFFMONETIZER_HTTP_PROXY=http://XXXXXXXXXX:XXXX
TRAFFMONETIZER_HTTPS_PROXY=http://XXXXXXXXXX:XXXX

## REPOCKET
REPOCKET_EMAIL=XXX@XXX.XXX
REPOCKET_PASSWD=XXXXXXXXyourREPOCKETPwyourREPOCKETPwDYXXX
REPOCKET_HTTP_PROXY=http://XXXXXXXXXX:XXXX
REPOCKET_HTTPS_PROXY=http://XXXXXXXXXX:XXXX

# BITPING
# Bitping have a special firt startup process , see manual setup documentation or use te setup script
BITPING_HTTP_PROXY=http://XXXXXXXXXX:XXXX
BITPING_HTTPS_PROXY=http://XXXXXXXXXX:XXXX

# WATCHTOWER
# use this container to auto-update containers and send notifications using shoutrrr
# SHOUTRRR_URL should be like this <app>://<token>@<webhook>
# REGISTER ON DISCORD AND CREATE A NEW SERVER, GO TO SERVER SETTINGS>INTEGRATIONS AND CREATE A WEBHOOK, THEN INSERT IT BELOW THE SAME FORMAT WRITTEN ABOVE. (YOU CAN ALSO USE OTHER APPS)
# SHOUTRRR_URL=yourApp:yourToken@yourWebHook

**Reply by MRColorR** [reply]
Thanks for this detailed explanation. I will be running a ubuntu vm and try to replicate it ASAP.

Keep this in mind about special characters in passwords and names - I could fix this on my side BUT there's a reason it's not already handled - some apps require "password/device shouldn't have special characters" and their Teams won't support or help you troubleshoot if your password contains special characters.

To avoid these problems you should use alphanumeric passwords for the accounts of all these apps.
This restriction is imposed upstream by the apps that are in the project and I can't change its behavior, which is why it wouldn't have made much sense to handle special characters that come out in scripts. Anyway I did some tests and for example characters like `-` seem to be handled correctly by all apps and also by project scripts. So if (rightly I would say) you don't want to give up the extra security and use a more complex password you can try using characters like -_. they should be safe to use.
I just did some testing using - so if you find any other characters that work let me know as I'm thinking of adding a disclaimer about valid characters for passwords at the top of the env config and in the wiki.

**Reply by claro-lukas** [reply]
Thank you for your reply.
We already tested passwords without special characters and thųs haven't helped.
The main problem is that proxies are not used and for example PEER2PROFIT in the devices list still show real computer IP instead of proxy IP.

**Reply by MRColorR** [reply]
Thank you for the info,  we are discussing a possible permanent fix using redsocks in #30 , this will require more resources but should ensure all apps work as it will be transparent to them. The price will be a little more resources occupied. 

As I already wrote, the solution is already almost entirely adaptable to our case and if someone wants to try to add the new solution and then add it to the project with a pull request I would be happy.
Otherwise I will work on it as soon as this first period of 2023 has passed, which is very intense for me.

**Comment by MRColorR** [comment, marked as answer]
The latest release of the project solves this problem 👍
